src/lib/utils/ct_utils.h

```diff
@@ -422,6 +428,87 @@ class Mask final {
       T m_mask;
 };
 
+/**
+* A CT::Option<T> is either a valid T, or not
+*
+* To maintain constant time behavior a value must always be stored.
+* A CT::Choice tracks is the value is valid or not. It is not possible
+* to access the inner value if the Choice is unset.
+*/
+template <typename T>
+class Option {
+   public:
+      /// Construct an Option which contains the specified value, and is set or not
+      constexpr Option(T v, Choice valid) : m_is_some(valid), m_value(std::move(v)) {}
+
+      /// Construct a set option with the provided value
+      constexpr Option(T v) : Option(std::move(v), Choice::yes()) {}
+
+      /// Construct an unset option with a default inner value
+      constexpr Option()
+         requires std::default_initializable<T>
+            : Option(T(), Choice::no()) {}
+
+      /// Return true if this Option contains a value
+      constexpr Choice has_value() const { return m_is_some; }
+
+      /**
+      * Apply a function to the inner value and return a new Option
+      * which contains that value. This is constant time only if @p f is.
+      *
+      * @note The function will always be called, even if the Option is None. It
+      *       must be prepared to handle any possible state of T.
+      */
+      template <std::invocable<const T&> F>
+      constexpr auto transform(F f) const -> Option<std::invoke_result_t<F, const T&>> {
+         return {f(m_value), m_is_some};
+      }
+
+      /// Either returns the value or throws an exception
+      T value() const {
+         BOTAN_STATE_CHECK(m_is_some.as_bool());
+         return m_value;
+      }
```

Comment:
Should that really return a copy of `m_value`? Maybe:

```suggestion
      /// Either returns the value or throws an exception
      const T& value() const {
         BOTAN_STATE_CHECK(m_is_some.as_bool());
         return m_value;
      }
```

... for completeness, [`std::optional` defines this both for lvalue and rvalue `this`](https://en.cppreference.com/w/cpp/utility/optional/value). Not sure we really need that, though. It just optimizes cases where a temporary is passed on.

```C++
takes_an_Option_by_value(returns_an_option_by_value().value())
```

Reply:
I suppose I was thinking of `Option` as largely for holding types that are cheap to copy (integers, simple arrays, etc) where this wouldn't matter, but yeah no reason not to return a reference here, `CT::Option<vector<uint8_t>>` is entirely sensible and might indeed be something worth using in PK decryption or the like.

Reply:
When we foresee to use it with `std::vector<uint8_t>` it would actually make sense to implement the lvalue/rvalue overloads that `std::optional<>` implements. In fact, this just made me re-think the `strong_type_unwrap()` helper: https://github.com/randombit/botan/pull/4170/files#r1672372281

Reply:
Reading https://stackoverflow.com/questions/24824432/what-is-use-of-the-ref-qualifier-const I am unconvinced by `const &&` and don't really see benefit. I'm going to merge this as is and if/when Option of non-trivally-copyable types comes up we can consider the best course.